### PR TITLE
refactor(wakunode2): added peer storage app setup proc and tests (pt. 2)

### PR DESCRIPTION
**Pull request by LNSD**
This is the second part of the `App` implementation and test cases.

> **Note**
> This PR is not merging into the `master` branch. It is merging into the `wakunode-develop` feature branch.

- [x] Added the setup logic for the JSON-RPC server, the REST server, and the metrics server.
- [x] Added shutdown logic for the JSON-RPC and REST servers.
- [x] Added a smoke test that asserts the correct initialization of the Wakunode2 App instance with the default configuration.
- [x] Enabled the `testwakunode2` tests (a.k.a. Wakunode2 App tests) in the CI pipeline.

### CODE REVIEW

File: apps/wakunode2/wakunode2.nim
Comment (LNSD) [inline apps/wakunode2/wakunode2.nim:1]: I had to rename the `node` variable to `nodeInstance` to solve the conflict of the `App`'s `node` getter.

File: tests/wakunode2/test_app.nim [hunk 1]
```diff
@@ -53,3 +56,29 @@ suite "Wakunode2 - App initialization":
 
     ## Then
     check res.isOk()
+
+  test "node setup is successful":
```
Comment (LNSD) [inline tests/wakunode2/test_app.nim:60]: I cannot use `asyncTest` here due to some non-GC-safe issue. I am using `waitFor` as a temporary workaround.

File: apps/wakunode2/wakunode2.nim
Comment (LNSD) [inline apps/wakunode2/wakunode2.nim:1]: I had to rename the `node` variable to `nodeInstance` to solve the conflict of the `App`'s `node` getter.

File: tests/wakunode2/test_app.nim [hunk 2]
```diff
@@ -53,3 +55,28 @@ suite "Wakunode2 - App initialization":
 
     ## Then
     check res.isOk()
+
+  test "node setup is successful with default configuration":
```
Comment (LNSD) [inline tests/wakunode2/test_app.nim:59]: I cannot use `asyncTest` here due to some non-GC-safe issue that makes the `wakunode2.setupWakuArchive()` incompatible with it. 

I am using `waitFor` as a temporary workaround.

File: apps/wakunode2/app.nim [hunk 3]
```diff
@@ -222,6 +226,41 @@ proc setupWakuArchiveDriver(dbUrl: string, vacuum: bool, migrate: bool): AppResu
...
+                                                migrate = app.conf.storeMessageDbMigration)
+  if archiveDriverRes.isOk():
+    app.archiveDriver = some(archiveDriverRes.get())
+  else:
+    return err("failed to configure archive driver" & archiveDriverRes.error)
+
+  # Message store retention policy
+  let storeMessageRetentionPolicyRes = validateStoreMessageRetentionPolicy(app.conf.storeMessageRetentionPolicy)
+  if storeMessageRetentionPolicyRes.isErr():
+    return err("failed to configure the message retention policy" & storeMessageRetentionPolicyRes.error)
```
Comment (Ivansete-status) [inline apps/wakunode2/app.nim:251]: The PR LGTM!
Shall we add a separator (e.g. colon or `" [" & error & "]"`) ?
There few other cases in this file that also need to separate errors.

**Review by Ivansete-status — APPROVED** [review]

File: apps/wakunode2/app.nim (hunk 3 above)
Comment (LNSD) [reply apps/wakunode2/app.nim]: Added. Note that these errors are here for backward compatibility and temporarily until the rest of the `App` is reworked and refactored.

File: apps/wakunode2/app.nim [hunk 4]
```diff
@@ -252,6 +291,17 @@ proc retrieveDynamicBootstrapNodes*(dnsDiscovery: bool, dnsDiscoveryUrl: string,
...
   ok(newSeq[RemotePeerInfo]()) # Return an empty seq by default
 
+proc setupDyamicBootstrapNodes*(app: var App): AppResult[void] =
+  let dynamicBootstrapNodesRes = retrieveDynamicBootstrapNodes(app.conf.dnsDiscovery,
+                                                               app.conf.dnsDiscoveryUrl,
+                                                               app.conf.dnsDiscoveryNameServers)
+  if dynamicBootstrapNodesRes.isOk():
+    app.dynamicBootstrapNodes = dynamicBootstrapNodesRes.get()
+  else:
+    warn "2/7 Retrieving dynamic bootstrap nodes failed. Continuing without dynamic bootstrap nodes.", error=dynamicBootstrapNodesRes.error
```
Comment (vpavlin) [inline apps/wakunode2/app.nim:301]: What is this 2/7?

**Comment by vpavlin** [comment]
Just a thought - any reason why we use `withXYZ` in node/builder and `setupXYZ` in the app? Would it make sense to keep this consistent?

File: apps/wakunode2/app.nim (hunk 4 above)
Comment (LNSD) [reply apps/wakunode2/app.nim:301]: Step 2 of 7. See:

https://github.com/waku-org/nwaku/blob/73cbafa658295213298b06ef7b349683f0d1f858/apps/wakunode2/wakunode2.nim#L719-L721

Comment (vpavlin) [reply apps/wakunode2/app.nim:301]: Ah, I see:D

**Comment by LNSD** [comment]
> Just a thought - any reason why we use `withXYZ` in node/builder and `setupXYZ` in the app? Would it make sense to keep this consistent?

Both things are different:

* The `WakuNodeBuilder` creates an instance of an embeddable `WakuNode`. This instance should not contain any logic related to the `wakunode2` infra node application.
* The `App`, when instantiated, on the other hand, would create a reference to the wakunode2 infra node application.

-------

Note that `WakuNode` and `wakunode2` are not the same.

The `wakunode2` app may include only relevant functionality if executed as a fleet infra node.
 
Opposed to that, the `waku/v2/waku_node#WakuNode` should be an embeddable node that contains the minimal functionality a node should perform; the idea is to provide the extra behavior to this minimal node via the builder methods (e.g., via Waku Relay handlers).

**Pull request merged**